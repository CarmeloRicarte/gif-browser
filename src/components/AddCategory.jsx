import { useState } from "react";
import PropTypes from "prop-types";
export const AddCategory = ({ onNewCategory }) => {
  const [inputValue, setInputValue] = useState("");

  /**
   * When the input changes, set the input value to the value of the input.
   */
  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };

  /**
   * The function onSubmit calls the function onNewCategory, which is passed in as a prop from the
   * parent component.
   * The function onNewCategory is passed the value of the input field.
   * @returns The input value is being returned.
   */
  const onSubmit = (event) => {
    event.preventDefault();
    const newInputValue = inputValue.trim();
    if (newInputValue.length <= 1) {
      return;
    }
    onNewCategory(newInputValue);
    setInputValue("");
  };

  return (
    <form aria-label="form" onSubmit={onSubmit}>
      <input
        type="text"
        id="seachGifs"
        placeholder="Buscar gifs"
        value={inputValue}
        onChange={onInputChange}
      />
    </form>
  );
};

AddCategory.propTypes = {
  onNewCategory: PropTypes.func.isRequired,
};
