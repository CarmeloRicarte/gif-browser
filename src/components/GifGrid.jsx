import { useEffect, useState } from "react";
import { getGifs } from "./../helpers/getGifs";
import { GifItem } from "./GifItem";

export const GifGrid = ({ category }) => {
  const [images, setImages] = useState([]);

  /**
   * GetImages is an async function that calls getGifs, which is an async function that returns a
   * promise, and then sets the state of images to the value of the promise
   */
  const getImages = async () => {
    const newImages = await getGifs(category);
    setImages(newImages);
  };

  useEffect(() => {
    getImages();
  }, []);

  return (
    <>
      <h3>{category}</h3>

      <div>
        {images.map((image) => (
          <GifItem key={image.id} {...image} />
        ))}
      </div>
    </>
  );
};
