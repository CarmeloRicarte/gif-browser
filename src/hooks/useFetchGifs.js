import { useEffect, useState } from "react";
import { getGifs } from "../helpers/getGifs";

/**
 * It returns an object with two properties: images and isLoading
 * @param category - The category of the gifs we want to fetch.
 * @returns An object with two properties: images and isLoading.
 */
export const useFetchGifs = (category) => {
    const [images, setImages] = useState([]);
    const [isLoading, setIsLoading] = useState(true);


    /**
     * An async function that gets the images from the getGifs function and sets the images to the
     * newImages.
     */
    const getImages = async () => {
        const newImages = await getGifs(category);
        setImages(newImages);
        setIsLoading(false);
    };

    useEffect(() => {
        getImages();
    }, []);

    return {
        images,
        isLoading
    }
}
