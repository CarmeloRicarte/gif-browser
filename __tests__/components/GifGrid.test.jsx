import { render, screen } from "@testing-library/react";
import { useFetchGifs } from "../../src/hooks/useFetchGifs";
import { GifGrid } from "./../../src/components/GifGrid";

vi.mock("../../src/hooks/useFetchGifs"); // do a mock of useFetchGifs
describe("Tests of <GifGrid />", () => {
  const category = "Ibai";

  test("should match with the snapshot", () => {
    useFetchGifs.mockReturnValue({
      images: [],
      isLoading: true,
    });
    const { container } = render(<GifGrid category={category} />);
    expect(container).toMatchSnapshot();
  });

  test("should show the loading at init", () => {
    useFetchGifs.mockReturnValue({
      images: [],
      isLoading: true,
    });
    render(<GifGrid category={category} />);
    expect(screen.getByText("Loading..."));
    expect(screen.getByText(category));
  });

  test("should show items when images are loaded with useFetchGifs", () => {
    const gifs = [
      {
        id: "1",
        url: "https://example.com/gifs/1.gif",
        title: "Ibai G2",
      },
      {
        id: "2",
        url: "https://example.com/gifs/2.gif",
        title: "Ibai de locos",
      },
    ];

    useFetchGifs.mockReturnValue({
      images: gifs,
      isLoading: false,
    });
    render(<GifGrid category={category} />);
    expect(screen.getAllByRole("img").length).toBe(2);
  });
});
