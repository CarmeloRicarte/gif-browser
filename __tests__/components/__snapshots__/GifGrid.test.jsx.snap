// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of <GifGrid /> should match with the snapshot 1`] = `
<div>
  <h3>
    Ibai
  </h3>
  <h4>
    Loading...
  </h4>
  <div
    class="card-grid"
  />
</div>
`;
